numpy/core/src/umath/ufunc_type_resolution.c

```diff
@@ -761,12 +761,24 @@ PyUFunc_AdditionTypeResolver(PyUFuncObject *ufunc,
     type_num2 = PyArray_DESCR(operands[1])->type_num;
 
     /* Use the default when datetime and timedelta are not involved */
-    if (!PyTypeNum_ISDATETIME(type_num1) && !PyTypeNum_ISDATETIME(type_num2)) {
+    if (!PyTypeNum_ISDATETIME(type_num1) && !PyTypeNum_ISDATETIME(type_num2)
+        && !(PyTypeNum_ISSTRING(type_num1) && PyTypeNum_ISSTRING(type_num2))) {
         return PyUFunc_SimpleUniformOperationTypeResolver(ufunc, casting,
                     operands, type_tup, out_dtypes);
     }
 
-    if (type_num1 == NPY_TIMEDELTA) {
+    if ((type_num1 == NPY_STRING && type_num2 == NPY_STRING)
+            || (type_num1 == NPY_UNICODE && type_num2 == NPY_UNICODE)) {
+        out_dtypes[0] = PyArray_PromoteTypes(PyArray_DESCR(operands[0]),
+                                             PyArray_DESCR(operands[1]));
+        if (out_dtypes[0] == NULL) {
+            return -1;
+        }
+        out_dtypes[1] = out_dtypes[0];
+        Py_INCREF(out_dtypes[1]);
+        out_dtypes[2] = out_dtypes[0];
+        Py_INCREF(out_dtypes[2]);
```

Comment:
Hmmmmmmmpf...  This needs at least a comment that it is wrong, but we don't care because the loop has the *correct* implementation and only the fact which string DType it is matters (and not e.g. the length).

TBH, I think you can just simplify it to `out_dtypes[0] = PyArray_DescrFromType(type_num1);` for everything with that same comment.  At least a zero sized dtype here is clearly nonsense and the promotion doesn't do anything *correct* anyway.